api/src/service/domain/organization/user_permission_grant.spec.ts

```diff
@@ -0,0 +1,112 @@
+const userId = "dummy";
+const baseUser: UserRecord = {
+  id: userId,
+  createdAt: new Date().toISOString(),
+  organization: "dummy",
+  displayName: "dummy",
+  passwordHash: "password",
+  permissions: {},
+  address: "12345",
+  encryptedPrivKey: "encrypted",
+  log: [],
+  additionalData: {},
+};
+
+const baseRepository = {
+  getUser: async () => baseUser,
+};
+
+describe("Granting user permissions: permissions", () => {
+  it("Without the user.intent.grantPermission permission, a user cannot grant user permissions", async () => {
+    const result = await grantUserPermission(ctx, alice, userId, bob.id, grantIntent, {
+      ...baseRepository,
+    });
+
+    // NotAuthorized error due to the missing permissions:
+    assert.isTrue(Result.isErr(result), "Alice is not authorized to grant this permission");
+    assert.instanceOf(result, NotAuthorized, "The error is of the type 'Not Authorized'");
+  });
+
+  it("The root user can always grant user permissions", async () => {
+    const result = await grantUserPermission(ctx, root, userId, bob.id, grantIntent, {
+      ...baseRepository,
+    });
+
+    assert.isTrue(Result.isOk(result));
+  });
+});
+
+describe("Granting user permissions: updates", () => {
+  it("The permission is granted if the user has the correct permissions", async () => {
+    const result = await grantUserPermission(ctx, alice, userId, bob.id, grantIntent, {
+      getUser: async () =>
+        Promise.resolve({
+          ...baseUser,
+          permissions: { "user.intent.grantPermission": [alice.id] },
+        }),
+    });
+    if (Result.isErr(result)) {
+      throw result;
+    }
+
+    const sourcedUser = result.reduce(
+      (user, event) => newUserFromEvent(ctx, user, event),
+      baseUser,
+    );
+    if (Result.isErr(sourcedUser)) {
+      throw sourcedUser;
+    }
+
+    assert.isTrue(Result.isOk(result), "Alice is authorized to grant this permission");
+    assert.isTrue(result.length > 0, "An event is created");
+  });
+
+  it("An existing permission is granted, nothing happens", async () => {
+    const testUser: UserRecord = {
+      id: userId,
+      createdAt: new Date().toISOString(),
+      organization: "dummy",
+      displayName: "dummy",
+      passwordHash: "password",
+      permissions: { "user.intent.grantPermission": [alice.id] },
+      address: "12345",
+      encryptedPrivKey: "encrypted",
+      log: [],
+      additionalData: {},
+    };
+    const result = await grantUserPermission(ctx, alice, userId, alice.id, grantIntent, {
+      getUser: async () => Promise.resolve(testUser),
+    });
+    if (Result.isErr(result)) {
+      throw result;
+    }
+
+    const sourcedUser = result.reduce(
+      (user, event) => newUserFromEvent(ctx, user, event),
+      testUser,
+    );
+    if (Result.isErr(sourcedUser)) {
+      throw sourcedUser;
+    }
+
+    assert.isTrue(Result.isOk(result), "Alice is authorized to grant this permission");
+    assert.deepEqual(result, []);
```

Comment:
in addition: Alice still has the permission

Reply:
implemented by using 
```js
// Alice still has the permission
    assert.isTrue(
      userAfterGrantingPermission.permissions["user.intent.grantPermission"]!.some(
        x => x === alice.id,
      ),
    );
```